pkg/codegen/operations.go

```diff
@@ -239,8 +242,43 @@ func (o *OperationDefinition) AllParams() []ParameterDefinition {
 // If we have parameters other than path parameters, they're bundled into an
 // object. Returns true if we have any of those. This is used from the template
 // engine.
+// A ParamsObject is only required when we do have at least a single non.reference type
+// parameter.
 func (o *OperationDefinition) RequiresParamObject() bool {
```

Comment:
This seems a bit fishy. The "ParamsObject" is the "params" structure passed into your handler in the case there are any parameters. Whether or not those parameters are references, we need this params object, since this is the container into which they are automatically unmarshaled. If we don't generate a params object when there are reference query params, we won't have a place to unmarshal them, and the handler code will be broken. Can you regenerate all boilerplate in the entire project and make sure that it works with your change?

Reply:
thanks for the feedback, found quite some problems after regenerating all of the test specs.
will ping you when I fixed them.

Reply:
I'm confused as to what you're trying to accomplish.

This ParamsObject needs to be generate if you have any query, header or cookie params, whether or not they are references. Path parameters go into the path and turn into function args.

Reply:
We do need to generate Models for any inline parameters, otherwise, we can't bind them, so whether or not something is a ref, we need to generate a Go struct for it.

Reply:
the old behavior generates three types when passed this spec:
```yaml
package: test
generate:
  models: true
  echo-server: true
output: parameters-ref-type.gen.go
```

```yaml
openapi: "3.0.2"
info:
  description: example openapi spec
  version: v1
  title: example openapi spec

servers:
  - url: "https://example.com/api/v1"
    description: Production

paths:
  /project:
    get:
      parameters:
        - $ref: "#/components/parameters/FilterQueryParam"
        - $ref: "#/components/parameters/SizeParam"
        - $ref: "#/components/parameters/OffsetParam"
        - $ref: "#/components/parameters/SortQueryParam"
      tags:
        - project
      summary: "Get list with all projects"
      description: "Get list with all projects."
      operationId: "getProjectList"
      responses:
        200:
          description: "List with projects"
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/ProjectList"

  /inventory:
    get:
      parameters:
        - $ref: "#/components/parameters/FilterQueryParam"
        - $ref: "#/components/parameters/SizeParam"
        - $ref: "#/components/parameters/OffsetParam"
        - $ref: "#/components/parameters/SortQueryParam"
      tags:
        - inventory
      summary: "Get list with all inventories"
      description: "Get list with all inventories."
      operationId: "getInventoryList"
      responses:
        200:
          description: "List with inventories"
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/InventoryList"

  /component:
    get:
      parameters:
        - in: query
          name: filterParam
          required: false
          schema:
            type: string
        - $ref: "#/components/parameters/SizeParam"
        - in: query
          name: offset
          required: false
          description: "The number of items to skip before starting to collect the result set"
          schema:
            type: integer
            minimum: 0
        - $ref: "#/components/parameters/SortQueryParam"
      tags:
        - component
      summary: "Get list with all components"
      description: "Get list with all components."
      operationId: "getComponentList"
      responses:
        200:
          description: "Request was successful"


components:
  schemas:
    ProjectList:
      type: array
      items:
        $ref: "#/components/schemas/ProjectListItem"

    ProjectListItem:
      type: object
      properties:
        id:
          type: string
          format: uuid
        name:
          type: string
        description:
          type: string
      required:
        - id
        - name

    InventoryList:
      type: array
      items:
        $ref: "#/components/schemas/InventoryListItem"

    InventoryListItem:
      type: object
      properties:
        id:
          type: string
          format: uuid
        project_id:
          type: string
          format: uuid
        name:
          type: string
        description:
          type: string
      required:
        - id
        - project_id
        - name

  parameters:
    SizeParam:
      in: query
      name: size
      required: false
      description: "Maximum number of items in result list"
      schema:
        type: integer
        minimum: -1
        default: 50

    OffsetParam:
      in: query
      name: offset
      required: false
      description: "The number of items to skip before starting to collect the result set"
      schema:
        type: integer
        minimum: 0

    FilterQueryParam:
      in: query
      name: filterParam
      required: false
      schema:
        type: string

    SortQueryParam:
      in: query
      name: sort
      schema:
        type: array
        items:
          type: string
```

old generated type code:
```go
// Package test provides primitives to interact with the openapi HTTP API.
//
// Code generated by github.com/deepmap/oapi-codegen version v1.11.0 DO NOT EDIT.
package test

import (
	"fmt"
	"net/http"

	"github.com/deepmap/oapi-codegen/pkg/runtime"
	openapi_types "github.com/deepmap/oapi-codegen/pkg/types"
	"github.com/labstack/echo/v4"
)

// InventoryList defines model for InventoryList.
type InventoryList = []InventoryListItem

// InventoryListItem defines model for InventoryListItem.
type InventoryListItem struct {
	Description *string            `json:"description,omitempty"`
	Id          openapi_types.UUID `json:"id"`
	Name        string             `json:"name"`
	ProjectId   openapi_types.UUID `json:"project_id"`
}

// ProjectList defines model for ProjectList.
type ProjectList = []ProjectListItem

// ProjectListItem defines model for ProjectListItem.
type ProjectListItem struct {
	Description *string            `json:"description,omitempty"`
	Id          openapi_types.UUID `json:"id"`
	Name        string             `json:"name"`
}

// FilterQueryParam defines model for FilterQueryParam.
type FilterQueryParam = string

// OffsetParam defines model for OffsetParam.
type OffsetParam = int

// SizeParam defines model for SizeParam.
type SizeParam = int

// SortQueryParam defines model for SortQueryParam.
type SortQueryParam = []string

// GetComponentListParams defines parameters for GetComponentList.
type GetComponentListParams struct {
	FilterParam *string `form:"filterParam,omitempty" json:"filterParam,omitempty"`

	// Maximum number of items in result list
	Size *SizeParam `form:"size,omitempty" json:"size,omitempty"`

	// The number of items to skip before starting to collect the result set
	Offset *int            `form:"offset,omitempty" json:"offset,omitempty"`
	Sort   *SortQueryParam `form:"sort,omitempty" json:"sort,omitempty"`
}

// GetInventoryListParams defines parameters for GetInventoryList.
type GetInventoryListParams struct {
	FilterParam *FilterQueryParam `form:"filterParam,omitempty" json:"filterParam,omitempty"`

	// Maximum number of items in result list
	Size *SizeParam `form:"size,omitempty" json:"size,omitempty"`

	// The number of items to skip before starting to collect the result set
	Offset *OffsetParam    `form:"offset,omitempty" json:"offset,omitempty"`
	Sort   *SortQueryParam `form:"sort,omitempty" json:"sort,omitempty"`
}

// GetProjectListParams defines parameters for GetProjectList.
type GetProjectListParams struct {
	FilterParam *FilterQueryParam `form:"filterParam,omitempty" json:"filterParam,omitempty"`

	// Maximum number of items in result list
	Size *SizeParam `form:"size,omitempty" json:"size,omitempty"`

	// The number of items to skip before starting to collect the result set
	Offset *OffsetParam    `form:"offset,omitempty" json:"offset,omitempty"`
	Sort   *SortQueryParam `form:"sort,omitempty" json:"sort,omitempty"`
}

// ServerInterface represents all server handlers.
type ServerInterface interface {
	// Get list with all components
	// (GET /component)
	GetComponentList(ctx echo.Context, params GetComponentListParams) error
	// Get list with all inventories
	// (GET /inventory)
	GetInventoryList(ctx echo.Context, params GetInventoryListParams) error
	// Get list with all projects
	// (GET /project)
	GetProjectList(ctx echo.Context, params GetProjectListParams) error
}

// ServerInterfaceWrapper converts echo contexts to parameters.
type ServerInterfaceWrapper struct {
	Handler ServerInterface
}

// GetComponentList converts echo context to params.
func (w *ServerInterfaceWrapper) GetComponentList(ctx echo.Context) error {
	var err error

	// Parameter object where we will unmarshal all parameters from the context
	var params GetComponentListParams
	// ------------- Optional query parameter "filterParam" -------------

	err = runtime.BindQueryParameter("form", true, false, "filterParam", ctx.QueryParams(), &params.FilterParam)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid format for parameter filterParam: %s", err))
	}

	// ------------- Optional query parameter "size" -------------

	err = runtime.BindQueryParameter("form", true, false, "size", ctx.QueryParams(), &params.Size)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid format for parameter size: %s", err))
	}

	// ------------- Optional query parameter "offset" -------------

	err = runtime.BindQueryParameter("form", true, false, "offset", ctx.QueryParams(), &params.Offset)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid format for parameter offset: %s", err))
	}

	// ------------- Optional query parameter "sort" -------------

	err = runtime.BindQueryParameter("form", true, false, "sort", ctx.QueryParams(), &params.Sort)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid format for parameter sort: %s", err))
	}

	// Invoke the callback with all the unmarshalled arguments
	err = w.Handler.GetComponentList(ctx, params)
	return err
}

// GetInventoryList converts echo context to params.
func (w *ServerInterfaceWrapper) GetInventoryList(ctx echo.Context) error {
	var err error

	// Parameter object where we will unmarshal all parameters from the context
	var params GetInventoryListParams
	// ------------- Optional query parameter "filterParam" -------------

	err = runtime.BindQueryParameter("form", true, false, "filterParam", ctx.QueryParams(), &params.FilterParam)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid format for parameter filterParam: %s", err))
	}

	// ------------- Optional query parameter "size" -------------

	err = runtime.BindQueryParameter("form", true, false, "size", ctx.QueryParams(), &params.Size)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid format for parameter size: %s", err))
	}

	// ------------- Optional query parameter "offset" -------------

	err = runtime.BindQueryParameter("form", true, false, "offset", ctx.QueryParams(), &params.Offset)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid format for parameter offset: %s", err))
	}

	// ------------- Optional query parameter "sort" -------------

	err = runtime.BindQueryParameter("form", true, false, "sort", ctx.QueryParams(), &params.Sort)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid format for parameter sort: %s", err))
	}

	// Invoke the callback with all the unmarshalled arguments
	err = w.Handler.GetInventoryList(ctx, params)
	return err
}

// GetProjectList converts echo context to params.
func (w *ServerInterfaceWrapper) GetProjectList(ctx echo.Context) error {
	var err error

	// Parameter object where we will unmarshal all parameters from the context
	var params GetProjectListParams
	// ------------- Optional query parameter "filterParam" -------------

	err = runtime.BindQueryParameter("form", true, false, "filterParam", ctx.QueryParams(), &params.FilterParam)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid format for parameter filterParam: %s", err))
	}

	// ------------- Optional query parameter "size" -------------

	err = runtime.BindQueryParameter("form", true, false, "size", ctx.QueryParams(), &params.Size)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid format for parameter size: %s", err))
	}

	// ------------- Optional query parameter "offset" -------------

	err = runtime.BindQueryParameter("form", true, false, "offset", ctx.QueryParams(), &params.Offset)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid format for parameter offset: %s", err))
	}

	// ------------- Optional query parameter "sort" -------------

	err = runtime.BindQueryParameter("form", true, false, "sort", ctx.QueryParams(), &params.Sort)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid format for parameter sort: %s", err))
	}

	// Invoke the callback with all the unmarshalled arguments
	err = w.Handler.GetProjectList(ctx, params)
	return err
}

// This is a simple interface which specifies echo.Route addition functions which
// are present on both echo.Echo and echo.Group, since we want to allow using
// either of them for path registration
type EchoRouter interface {
	CONNECT(path string, h echo.HandlerFunc, m ...echo.MiddlewareFunc) *echo.Route
	DELETE(path string, h echo.HandlerFunc, m ...echo.MiddlewareFunc) *echo.Route
	GET(path string, h echo.HandlerFunc, m ...echo.MiddlewareFunc) *echo.Route
	HEAD(path string, h echo.HandlerFunc, m ...echo.MiddlewareFunc) *echo.Route
	OPTIONS(path string, h echo.HandlerFunc, m ...echo.MiddlewareFunc) *echo.Route
	PATCH(path string, h echo.HandlerFunc, m ...echo.MiddlewareFunc) *echo.Route
	POST(path string, h echo.HandlerFunc, m ...echo.MiddlewareFunc) *echo.Route
	PUT(path string, h echo.HandlerFunc, m ...echo.MiddlewareFunc) *echo.Route
	TRACE(path string, h echo.HandlerFunc, m ...echo.MiddlewareFunc) *echo.Route
}

// RegisterHandlers adds each server route to the EchoRouter.
func RegisterHandlers(router EchoRouter, si ServerInterface) {
	RegisterHandlersWithBaseURL(router, si, "")
}

// Registers handlers, and prepends BaseURL to the paths, so that the paths
// can be served under a prefix.
func RegisterHandlersWithBaseURL(router EchoRouter, si ServerInterface, baseURL string) {

	wrapper := ServerInterfaceWrapper{
		Handler: si,
	}

	router.GET(baseURL+"/component", wrapper.GetComponentList)
	router.GET(baseURL+"/inventory", wrapper.GetInventoryList)
	router.GET(baseURL+"/project", wrapper.GetProjectList)

}
```


the new updated behavior only produces a single type for each globally defined parameter:

```go
// Package test provides primitives to interact with the openapi HTTP API.
//
// Code generated by unknown module path version unknown version DO NOT EDIT.
package test

import (
	"fmt"
	"net/http"

	"github.com/deepmap/oapi-codegen/pkg/runtime"
	openapi_types "github.com/deepmap/oapi-codegen/pkg/types"
	"github.com/labstack/echo/v4"
)

// InventoryList defines model for InventoryList.
type InventoryList = []InventoryListItem

// InventoryListItem defines model for InventoryListItem.
type InventoryListItem struct {
	Description *string            `json:"description,omitempty"`
	Id          openapi_types.UUID `json:"id"`
	Name        string             `json:"name"`
	ProjectId   openapi_types.UUID `json:"project_id"`
}

// ProjectList defines model for ProjectList.
type ProjectList = []ProjectListItem

// ProjectListItem defines model for ProjectListItem.
type ProjectListItem struct {
	Description *string            `json:"description,omitempty"`
	Id          openapi_types.UUID `json:"id"`
	Name        string             `json:"name"`
}

// FilterQueryParam defines model for FilterQueryParam.
type FilterQueryParam = string

// OffsetParam defines model for OffsetParam.
type OffsetParam = int

// SizeParam defines model for SizeParam.
type SizeParam = int

// SortQueryParam defines model for SortQueryParam.
type SortQueryParam = []string

// GetComponentListParams defines parameters for GetComponentList.
type GetComponentListParams struct {
	FilterParam *string `form:"filterParam,omitempty" json:"filterParam,omitempty"`

	// The number of items to skip before starting to collect the result set
	Offset *int `form:"offset,omitempty" json:"offset,omitempty"`
}

// ServerInterface represents all server handlers.
type ServerInterface interface {
	// Get list with all components
	// (GET /component)
	GetComponentList(ctx echo.Context, params GetComponentListParams, size SizeParam, sort SortQueryParam) error
	// Get list with all inventories
	// (GET /inventory)
	GetInventoryList(ctx echo.Context, filterParam FilterQueryParam, size SizeParam, offset OffsetParam, sort SortQueryParam) error
	// Get list with all projects
	// (GET /project)
	GetProjectList(ctx echo.Context, filterParam FilterQueryParam, size SizeParam, offset OffsetParam, sort SortQueryParam) error
}

// ServerInterfaceWrapper converts echo contexts to parameters.
type ServerInterfaceWrapper struct {
	Handler ServerInterface
}

// GetComponentList converts echo context to params.
func (w *ServerInterfaceWrapper) GetComponentList(ctx echo.Context) error {
	var err error

	// Parameter object where we will unmarshal all parameters from the context
	var params GetComponentListParams

	// ------------- Optional query parameter "filterParam" -------------

	err = runtime.BindQueryParameter("form", true, false, "filterParam", ctx.QueryParams(), &params.FilterParam)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid format for parameter filterParam: %s", err))
	}

	// ------------- Optional query parameter "size" -------------

	var size SizeParam

	err = runtime.BindQueryParameter("form", true, false, "size", ctx.QueryParams(), &size)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid format for parameter size: %s", err))
	}

	// ------------- Optional query parameter "offset" -------------

	err = runtime.BindQueryParameter("form", true, false, "offset", ctx.QueryParams(), &params.Offset)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid format for parameter offset: %s", err))
	}

	// ------------- Optional query parameter "sort" -------------

	var sort SortQueryParam

	err = runtime.BindQueryParameter("form", true, false, "sort", ctx.QueryParams(), &sort)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid format for parameter sort: %s", err))
	}

	// Invoke the callback with all the unmarshalled arguments
	err = w.Handler.GetComponentList(ctx, params, size, sort)
	return err
}

// GetInventoryList converts echo context to params.
func (w *ServerInterfaceWrapper) GetInventoryList(ctx echo.Context) error {
	var err error

	// ------------- Optional query parameter "filterParam" -------------

	var filterParam FilterQueryParam

	err = runtime.BindQueryParameter("form", true, false, "filterParam", ctx.QueryParams(), &filterParam)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid format for parameter filterParam: %s", err))
	}

	// ------------- Optional query parameter "size" -------------

	var size SizeParam

	err = runtime.BindQueryParameter("form", true, false, "size", ctx.QueryParams(), &size)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid format for parameter size: %s", err))
	}

	// ------------- Optional query parameter "offset" -------------

	var offset OffsetParam

	err = runtime.BindQueryParameter("form", true, false, "offset", ctx.QueryParams(), &offset)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid format for parameter offset: %s", err))
	}

	// ------------- Optional query parameter "sort" -------------

	var sort SortQueryParam

	err = runtime.BindQueryParameter("form", true, false, "sort", ctx.QueryParams(), &sort)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid format for parameter sort: %s", err))
	}

	// Invoke the callback with all the unmarshalled arguments
	err = w.Handler.GetInventoryList(ctx, filterParam, size, offset, sort)
	return err
}

// GetProjectList converts echo context to params.
func (w *ServerInterfaceWrapper) GetProjectList(ctx echo.Context) error {
	var err error

	// ------------- Optional query parameter "filterParam" -------------

	var filterParam FilterQueryParam

	err = runtime.BindQueryParameter("form", true, false, "filterParam", ctx.QueryParams(), &filterParam)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid format for parameter filterParam: %s", err))
	}

	// ------------- Optional query parameter "size" -------------

	var size SizeParam

	err = runtime.BindQueryParameter("form", true, false, "size", ctx.QueryParams(), &size)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid format for parameter size: %s", err))
	}

	// ------------- Optional query parameter "offset" -------------

	var offset OffsetParam

	err = runtime.BindQueryParameter("form", true, false, "offset", ctx.QueryParams(), &offset)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid format for parameter offset: %s", err))
	}

	// ------------- Optional query parameter "sort" -------------

	var sort SortQueryParam

	err = runtime.BindQueryParameter("form", true, false, "sort", ctx.QueryParams(), &sort)
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid format for parameter sort: %s", err))
	}

	// Invoke the callback with all the unmarshalled arguments
	err = w.Handler.GetProjectList(ctx, filterParam, size, offset, sort)
	return err
}

// This is a simple interface which specifies echo.Route addition functions which
// are present on both echo.Echo and echo.Group, since we want to allow using
// either of them for path registration
type EchoRouter interface {
	CONNECT(path string, h echo.HandlerFunc, m ...echo.MiddlewareFunc) *echo.Route
	DELETE(path string, h echo.HandlerFunc, m ...echo.MiddlewareFunc) *echo.Route
	GET(path string, h echo.HandlerFunc, m ...echo.MiddlewareFunc) *echo.Route
	HEAD(path string, h echo.HandlerFunc, m ...echo.MiddlewareFunc) *echo.Route
	OPTIONS(path string, h echo.HandlerFunc, m ...echo.MiddlewareFunc) *echo.Route
	PATCH(path string, h echo.HandlerFunc, m ...echo.MiddlewareFunc) *echo.Route
	POST(path string, h echo.HandlerFunc, m ...echo.MiddlewareFunc) *echo.Route
	PUT(path string, h echo.HandlerFunc, m ...echo.MiddlewareFunc) *echo.Route
	TRACE(path string, h echo.HandlerFunc, m ...echo.MiddlewareFunc) *echo.Route
}

// RegisterHandlers adds each server route to the EchoRouter.
func RegisterHandlers(router EchoRouter, si ServerInterface) {
	RegisterHandlersWithBaseURL(router, si, "")
}

// Registers handlers, and prepends BaseURL to the paths, so that the paths
// can be served under a prefix.
func RegisterHandlersWithBaseURL(router EchoRouter, si ServerInterface, baseURL string) {

	wrapper := ServerInterfaceWrapper{
		Handler: si,
	}

	router.GET(baseURL+"/component", wrapper.GetComponentList)
	router.GET(baseURL+"/inventory", wrapper.GetInventoryList)
	router.GET(baseURL+"/project", wrapper.GetProjectList)

}
```

Reply:
This additional type is generated with this update for all of the inline parameters that are not referencing global types in the spec.
```go
// GetComponentListParams defines parameters for GetComponentList.
type GetComponentListParams struct {
	FilterParam *string `form:"filterParam,omitempty" json:"filterParam,omitempty"`

	// The number of items to skip before starting to collect the result set
	Offset *int `form:"offset,omitempty" json:"offset,omitempty"`
}
```